### DISCUSSION TITLE
Reinforcement learning?

### DISCUSSION THREAD

**Discussion by nfreesto** [Kotlin/kotlindl#46] [Q&A] [answered]
hey, I'm pretty new to AI and Neural Networks as a whole, so excuse me if I'm incorrect, but it seems like there is no currently implemented way to train a model using reinforcement learning. Is there a way to achieve this with currently implemented methods? Barring that, is that something that's in the plans for the future of this library?

Thanks for your help

**Comment by zaleslaw** [comment, marked as answer]
Hi, @NahatanIII thanks for the raised question. You are not the first person who is interested in RL with Kotlin.
Unfortunately, the current models are related to the classic Computer Vision domain and implemented such CNN architectures like VGG, ResNet, and MobileNet (I mean upcoming release 0.2)

But we are not limit this library to mainstream Deep Learning only and the RL package is possible in the future if it will have high users demand.

Could I ask you, what the basic RL models and for what goals or in what fields are you going to apply these models? For playing video games, for solving optimization tasks, or something else? Please, share more about your needs.

**Reply by d-lowl** [reply]
I'm actually potentially interested in RL with KotlinDL myself (for a pet-project of mine), and might try sketching something up in the following weeks (months? i don't know, my schedule is a mess as usual). 

As @zaleslaw said, it'd be good if you could provide examples of models/features that you would like to have @NahatanIII. Maybe it'll be useful to work around those, rather than mine (I don't think that my pet's use-cases will be general enough to be useful for the public)

**Reply by dosier** [reply]
When I look at some of the [keras examples of RL models](https://keras.io/examples/rl/), I see that we have the layers used to build these models implemented in KotlinDL (like Conv2D). Maybe I am missing something but if KotlinDL has the layers u need, it should be achievable right (at least for these examples I mentioned)?

It would be nice to have the [keras examples of RL models](https://keras.io/examples/rl/) converted to KotlinDL and maybe some nice api could be build around them.

**Reply by d-lowl** [reply]
Yeah, I've had a look at those examples. The first obviously missing things are the gyms themselves (e.g. see the setup in [cartpole example](https://keras.io/examples/rl/actor_critic_cartpole/). Also I'm not sure if KotlinDL exposes all of the required API, that's to be found out, at least for me. Wrapping some things up with high level functions would also be useful I think (these examples use quite a bit of TF API directly)

**Reply by dosier** [reply]
True, maybe an issue could be made for implementing these gyms then? If it turns our KotlinDL does not expose all of the required API it'd be good to know.

**Reply by d-lowl** [reply]
Alright, I had a closer look. It looks like all of the examples use [OpenAI gym](https://gym.openai.com/docs/) which only has a python implementation (it used to have a REST API server implementation with bindings for a few languages, but it has been abandoned a few years ago). I don't think the gyms themselves belong to this repo. But I think I can implement one of them (CartPole looks relatively easy to tackle) in a separate repo and report back.

**Reply by d-lowl** [reply]
Upd: I cooked this up in a few hours https://github.com/d-lowl/kgym 
No proper unit tests, but it seems to work. Next thing is to implement that CartPole example with KotlinDL

**Reply by dosier** [reply]
Nice work :D! One other thing I just noticed is that KotlinDL doesn't have [GradientTape](https://www.tensorflow.org/api_docs/python/tf/GradientTape ) implemented. So maybe that should be a priority, once implemented and combined with your gym code, the full [Actor-Critic example](https://keras.io/examples/rl/actor_critic_cartpole/) from keras should be implementable in KotlinDL right?

**Reply by d-lowl** [reply]
Right, yeah, it looks like, again, tf java api  lacks GradientTape indeed.

**Reply by zaleslaw** [reply]
So, it is a great idea to start from the CartPole examples, but it will be hard without GradientType and low-level control of gradients. In my humble opinion, classic Keras is not designed well for the RL tasks.

But I want to share some ideas what could be starting points here:

1. Simple RL tasks could be implemented from scratch in Kotlin without TF backend usage; the simplest backpropagation for Dense layers could be added 
2. Implement something that could be added later to KotlinDL without TF backend usage at all: Hidden Markov chains for example.
2. [Train RL models in Python ](https://stackoverflow.com/questions/66576469/how-to-use-a-trained-rl-model-to-make-a-prediction)and try to use them for prediction in KotlinDL (under InferenceModel), for example
3. Wait for [GradientTape support](https://github.com/tensorflow/java/issues/285) in TF Java API 2.x and KotlinDL integration with this API (I think it's 2022)
4. Wait for (or even collaborate on) basic NN implementation in pure Kotlin in KotlinDL during 0.4-0.5 releases.
5. Wait for PyTorch integration during 0.5-0.6 releases and use an alternative backend with full control on the training cycle
6. Actively prepare examples in KotlinDL excluding backprop implementation
7. Learn this [repository](https://github.com/wumo/Reinforcement-Learning-An-Introduction) with Kotlin examples to Satton book or [this](https://github.com/chen0040/java-reinforcement-learning)

**Reply by d-lowl** [reply]
Good points.

1/ Is sensible, but probably isn't enough for my application. (Might be totally good enough for many other uses though)
4/ I think is the way to go, so not to reimplement the wheel (i.e. the GradientTape) -- but point 3/ is probably what I'll end up doing
8/ I like these resources! From my side, I can also recommend Stanford's course on RL, for theoretical background (if anyone here is seeking that), available on youtube: https://www.youtube.com/watch?v=FgzM3zpZ55o&list=PLoROMvodv4rOSOPzutgyCTapiGlY2Nd8u

**Reply by zaleslaw** [reply]
I hope about GradientTape too, now there are a few problems with low-level C/C++ API for that, but the TF core team works on it.

**Comment by lukasgppl** [comment]
I'm also interested in deep reinforcement learning with KotlinDL. 

Is there a possiblity by now, that allows me to implement a standard PPO for example? I'm still learning about KotlinDL, but I don't really see a way how I could implement the PPO objective. Is there a way to feed the calculated advantages into a custom loss function  that includes the clip in every episode? Any workarounds (like only using one epoch in the training with .fit)? 

Also, it would be crucial to export and import trained models. Is there a good way to do this? saveModel() and loadModel() don't seem like they are made for this purpose. 

I'd love to contribute to a simple drl example if I can get some assistance. I already wrote some code for it but I'd need some help with some details @zaleslaw.
